File: Gobchat.App/Module/Language/Internal/LocaleManager.cs
/*******************************************************************************
 * Copyright (C) 2019-2020 MarbleBag
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 *******************************************************************************/

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using Gobchat.Core.Resource;
using Gobchat.Core.Runtime;

namespace Gobchat.Module.Language.Internal
{
    internal sealed class LocaleManager : ILocaleManager
    {
        private static readonly NLog.Logger logger = NLog.LogManager.GetCurrentClassLogger();

        private event EventHandler<LocaleEventArgs> _localeChange;

        private LocalFolderResourceResolver _resourceResolver;
        private HjsonResourceLoader _resourceLoader;

        private object _lock = new object();

        public LocaleManager()
        {
            var languagePath = System.IO.Path.Combine(GobchatContext.ResourceLocation, @"lang");
            _resourceResolver = new LocalFolderResourceResolver(languagePath);
            _resourceLoader = new HjsonResourceLoader();

            ActiveCulture = DefaultCulture;
        }

        public CultureInfo DefaultCulture { get => CultureInfo.GetCultureInfo("en"); }

        public CultureInfo ActiveCulture { get; private set; }

        public event EventHandler<LocaleEventArgs> OnLocaleChange
        {
            add
            {
                lock (_lock)
                {
                    _localeChange += value;
                    value(this, new LocaleEventArgs(ActiveCulture));
                }
            }
            remove
            {
                _localeChange -= value;
            }
        }

        public string GetLocale(CultureInfo cultureInfo)
        {
            return cultureInfo.Name;
        }

        public void SwitchCulture(CultureInfo cultureInfo)
        {
            if (cultureInfo == null)
                cultureInfo = DefaultCulture;

            lock (_lock)
            {
                if (ActiveCulture.Equals(cultureInfo))
                    return;
                ActiveCulture = cultureInfo;
            }

            logger.Info($"Set locale to {ActiveCulture}");
            _localeChange?.Invoke(this, new LocaleEventArgs(ActiveCulture));
        }

        public IResourceBundle GetResourceBundle(string filename)
        {
            //TODO cache, auto update, selective resource loader depending on resource type
            var bundle = new CachedResourceBundle(_resourceResolver, _resourceLoader, filename, GetLocale(DefaultCulture));
            bundle.SetLocale(GetLocale(ActiveCulture));
            return bundle;
        }

        public R Build<R, M>() where R : MessageRegistry<M>
        {
            var registryType = typeof(R);
            var emptyConstructor2 = registryType.GetConstructor(Type.EmptyTypes);
            var registry = (R)emptyConstructor2.Invoke(Array.Empty<object>());

            Update(registry);
            return registry;
        }

        public IMessageRegistry<M> BuildMessageRegistry<M>()
        {
            var registry = new MessageRegistry<M>();
            Update(registry);
            return registry;
        }

        private void Update<M>(MessageRegistry<M> registry)
        {
            var messageType = typeof(M);
            var emptyConstructor = messageType.GetConstructor(Type.EmptyTypes);
            var message = (M)emptyConstructor.Invoke(Array.Empty<object>());
            registry.UpdateMessages(message);
        }
    }
}

File: Gobchat.App/resources/ui/gobchat/deprecated/LocaleManager.js
/*******************************************************************************
 * Copyright (C) 2019-2025 MarbleBag
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 *******************************************************************************/

'use strict'

var Gobchat = (function (Gobchat) {
    const DataAttributeText = "data-gob-locale-text"
    const DataAttributeTitle = "data-gob-locale-title"
    const LocaleClass = "gob-localized"
    const ElementSelector = `[${DataAttributeText}],[${DataAttributeTitle}]`

    async function updateLocaleOnElements(htmlElement, locale) {
        const selectedElements = $(htmlElement).find(ElementSelector).addBack(ElementSelector)
        if (selectedElements.length == 0)
            return

        const stringIds = []

        selectedElements.each(function () {
            const $this = $(this)
            if ($this.attr(DataAttributeText)) {
                const textId = $this.attr(DataAttributeText)
               // const tooltipId = `${textId}.tooltip` // needs to be streamlined to autodetect inputs and annotate them with the tooltip class
                stringIds.push(textId)
               // stringIds.push(tooltipId)
            }

            if ($this.attr(DataAttributeTitle))
                stringIds.push($this.attr(DataAttributeTitle))
        })

        if (stringIds.length == 0)
            return

        const lookup = await GobchatAPI.getLocalizedStrings(locale, stringIds)

        selectedElements.each(function () {
            const $this = $(this)

            if ($this.attr(DataAttributeText)) {
                const id = $this.attr(DataAttributeText)
                const txt = lookup[id]
                $this.html(txt)
            }

            if ($this.attr(DataAttributeTitle)) {
                const id = $this.attr(DataAttributeTitle)
                const txt = lookup[id]
                $this.prop("title", txt);
            }
        })
    }

    class LocaleManager {
        setLocale(locale) {
            this._locale = locale
        }

        async load(keys, params, lookup, language) {
            keys = [].concat(keys)
            language = language || this._locale

            var missingKeys = keys
            if (lookup)
                missingKeys = keys.filter(e => !(lookup[e] === undefined || lookup[e] === null))

            if (missingKeys.length > 0) {
                const localization = await GobchatAPI.getLocalizedStrings(language, missingKeys)
                if (lookup)
                    Object.entries(localization).forEach((k, v) => lookup[k] = v)
                else
                    lookup = localization
            }

            var results = lookup
            if (Object.keys(lookup).length !== keys.length)
                results = keys.map(e => lookup[e])

            if (params && params.length > 0)
                for (var key in Object.keys(results))
                    results[key] = Gobchat.formatString(results[key], params)

            if (keys.length === 1)
                return results[keys[0]]

            return results
        }

        async get(key, language) {
            const lookup = await this.getAll([key], language)
            return lookup[key]
        }

        async getAll(keys, language) {
            const lookup = await GobchatAPI.getLocalizedStrings(language || this._locale, keys || [])
            return lookup
        }

        async getAndFormat(key, params) {
            let value = await this.get(key)
            value = Gobchat.formatString(value, params)
            return value
        }

        async updateElement(element, language) {
            await updateLocaleOnElements(element, language || this._locale)
        }
    }

    Gobchat.LocaleManager = LocaleManager

    return Gobchat
}(Gobchat || {}));